wagtail/contrib/redirects/views.py

```diff
@@ -126,3 +133,183 @@ def add(request):
+            "form": ConfirmImportForm(dataset.headers, initial=initial),
+            "dataset": dataset,
+        },
+    )
+
+
+@permission_checker.require_any("add")
+@require_http_methods(["POST"])
+def process_import(request):
+    from_encoding = "utf-8"
+    form_kwargs = {}
+    form = ConfirmImportForm(
+        DEFAULT_FORMATS, request.POST or None, request.FILES or None, **form_kwargs
+    )
+
+    is_confirm_form_valid = form.is_valid()
+
+    import_formats = get_import_formats()
+    input_format = import_formats[int(form.cleaned_data["input_format"])]()
+
+    FileStorage = get_file_storage()
+    file_storage = FileStorage(name=form.cleaned_data["import_file_name"])
+
+    if not is_confirm_form_valid:
+        data = file_storage.read(input_format.get_read_mode())
+        if not input_format.is_binary() and from_encoding:
+            data = force_str(data, from_encoding)
+        dataset = input_format.create_dataset(data)
+
+        initial = {
+            "import_file_name": file_storage.name,
+            "original_file_name": form.cleaned_data["import_file_name"],
+            "input_format": form.cleaned_data["input_format"],
+        }
+
+        return render(
+            request,
+            "wagtailredirects/confirm_import.html",
+            {
+                "form": ConfirmImportForm(
+                    dataset.headers,
+                    request.POST or None,
+                    request.FILES or None,
+                    initial=initial,
+                ),
+                "dataset": dataset,
+            },
+        )
+
+    data = file_storage.read(input_format.get_read_mode())
+    if not input_format.is_binary() and from_encoding:
+        data = force_str(data, from_encoding)
+    dataset = input_format.create_dataset(data)
+
+    import_summary = create_redirects_from_dataset(
+        dataset,
+        {
+            "from_index": int(form.cleaned_data["from_index"]),
+            "to_index": int(form.cleaned_data["to_index"]),
+            "permanent": form.cleaned_data["permanent"],
+            "site": form.cleaned_data["site"],
+        },
+    )
+
+    file_storage.remove()
+
+    return render(
+        request,
+        "wagtailredirects/import_summary.html",
+        {
+            "form": ImportForm(DEFAULT_FORMATS),
+            "import_summary": import_summary,
+        },
+    )
+
+
+def create_redirects_from_dataset(dataset, config):
+    errors = []
+    successes = 0
+    total = 0
+
+    for row in dataset:
+        total += 1
+
+        from_link = row[config["from_index"]]
+        to_link = row[config["to_index"]]
+
+        data = {
+            "old_path": from_link,
+            "redirect_link": to_link,
+            "is_permanent": config["permanent"],
+        }
+
+        if config["site"]:
+            data["site"] = config["site"].pk
+
+        form = RedirectForm(data)
+        if not form.is_valid():
+            error = form.errors.as_text().replace("\n", "")
```

Comment:
These render as a row of errors with the `*` character in front.
e.g.
> * redirect_link * Enter a valid URL.* __all__ * A redirect with this path already exists.

It would be great if this could be slightly nicer to read, maybe pull out the leading `*` and join with a comma. so the output would be more like.
> redirect_link, Enter a valid URL., __all__, A redirect with this path already exists.

More could probably done to avoid these being too cryptic but I think this is much better than saying nothing.

Reply:
@lb- Good point, here I hijack the form validation and it's not always pretty to the user (especially the `__all__`key bit). Will apply some string parsing to enhance user experience.